File: internal/runtime/lua/player_composite_test.go
package modruntime

import (
	"context"
	"os"
	"testing"

	"github.com/gravestench/dark-magic/internal/content"
	"github.com/gravestench/dark-magic/internal/presentation/render"
	lua "github.com/yuin/gopher-lua"
)

func TestPlayerCompositeResolvesCOFLayerWeaponClasses(t *testing.T) {
	runtime := New()
	if err := runtime.RegisterInstaller(ContentRequire(content.D2Legacy(), "lua")); err != nil {
		t.Fatal(err)
	}
	if err := runtime.RegisterModule(Module{Name: "engine.render/v1", Loader: func(state *lua.LState) int {
		module := state.NewTable()
		state.SetField(module, "cof_info", state.NewFunction(func(state *lua.LState) int {
			got := state.CheckString(1)
			if got != "data/global/chars/AM/COF/AMWLHTH.cof" && got != "data/global/chars/AM/COF/AMWL1HS.cof" {
				state.RaiseError("unexpected COF %q", got)
			}
			info := state.NewTable()
			info.RawSetString("directions", lua.LNumber(16))
			layers := state.NewTable()
			for _, layer := range []struct{ component, weaponClass string }{{"HD", "1ht"}, {"RA", "hth"}, {"RH", "hth"}} {
				entry := state.NewTable()
				entry.RawSetString("type", lua.LString(layer.component))
				entry.RawSetString("weapon_class", lua.LString(layer.weaponClass))
				layers.Append(entry)
			}
			info.RawSetString("layers", layers)
			state.Push(info)
			return 1
		}))
		state.SetField(module, "asset_exists", state.NewFunction(func(state *lua.LState) int {
			path := state.CheckString(1)
			state.Push(lua.LBool(path == "data/global/chars/AM/HD/AMHDLITWL1HT.dcc" || path == "data/global/chars/AM/RA/AMRALITWLHTH.dcc"))
			return 1
		}))
		state.SetField(module, "animdata_info", state.NewFunction(func(state *lua.LState) int {
			got := state.CheckString(1)
			if got != "AMWLHTH" && got != "AMWL1HS" {
				state.RaiseError("unexpected animation key %q", got)
			}
			info := state.NewTable()
			info.RawSetString("speed", lua.LNumber(333))
			info.RawSetString("frames", lua.LNumber(8))
			info.RawSetString("events", state.NewTable())
			state.Push(info)
			return 1
		}))
		state.Push(module)
		return 1
	}}); err != nil {
		t.Fatal(err)
	}
	if err := runtime.Start(context.Background()); err != nil {
		t.Fatal(err)
	}
	defer runtime.Stop(context.Background())

	script := `
local composite=require("d2legacy.gameplay.player_composite").unarmed({
  token="AM", mode="WL", weapon_class="HTH", palette="data/global/Palette/units/pal.dat", direction=3,
})
assert(string.sub(composite.key,1,12)=="AM:WL:HTH:14")
assert(composite.direction==14 and composite.dcc_direction==3)
assert(composite.components.HD=="data/global/chars/AM/HD/AMHDLITWL1HT.dcc")
assert(composite.components.RA=="data/global/chars/AM/RA/AMRALITWLHTH.dcc")
assert(composite.components.RH==nil)
assert(composite.rate==333 and composite.frames==8)
local fine_direction=require("d2legacy.gameplay.player_composite").unarmed({
  token="AM", mode="WL", weapon_class="HTH", palette="data/global/Palette/units/pal.dat", direction=15,
})
assert(fine_direction.direction==15)
local equipped=require("d2legacy.gameplay.player_composite").resolve({
  token="AM", mode="WL", weapon_class="HTH", palette="data/global/Palette/units/pal.dat", direction=3,
},{active_weapon_set=0,items={{container="equipment",slot="rarm",weapon_set=0,weapon_class="1hs",composite={RH="ssd"}}}})
assert(equipped.cof=="data/global/chars/AM/COF/AMWL1HS.cof")
assert(equipped.components.RH=="data/global/chars/AM/RH/AMRHSSDWLHTH.dcc")
local playback=require("d2legacy.gameplay.player_composite").new_playback({mode="A"})
local crossed=require("d2legacy.gameplay.player_composite").advance(playback,{rate=256,frames=4,events={[2]=1,[3]=3},mode="A"},0.09)
assert(playback.frame==3 and #crossed==2 and crossed[1].event==1 and crossed[2].event==3)
`
	scope := &Scope{}
	if err := runtime.RunScoped(context.Background(), scope, func(state *lua.LState) error { return state.DoString(script) }); err != nil {
		t.Fatal(err)
	}
	if err := scope.Close(); err != nil {
		t.Fatal(err)
	}
}

func TestRealArchivesComposeUnarmedPlayerModes(t *testing.T) {
	directory := os.Getenv("DARK_MAGIC_TEST_MPQ_DIRECTORY")
	if directory == "" {
		t.Skip("set DARK_MAGIC_TEST_MPQ_DIRECTORY to the Diablo II MPQ directory")
	}
	t.Setenv("MPQ_DIRECTORY", directory)
	assets, err := content.FromEnvironment()
	if err != nil {
		t.Fatal(err)
	}

	runtime := New()
	var composer render.Composer
	if err := runtime.RegisterInstaller(ContentRequire(content.D2Legacy(), "lua")); err != nil {
		t.Fatal(err)
	}
	capability := NewRenderCapability(runtime, &composer, assets)
	if err := runtime.RegisterModule(capability.Module()); err != nil {
		t.Fatal(err)
	}
	if err := runtime.Start(context.Background()); err != nil {
		t.Fatal(err)
	}
	defer runtime.Stop(context.Background())

	script := `
local render=require("engine.render/v1")
local adapter=require("d2legacy.gameplay.player_composite")
for _,token in ipairs({"AM","SO","NE","PA","BA","AI","DZ"}) do
  for _,mode in ipairs({"NU","WL","RN"}) do
    local ok,err=pcall(function()
      local composite=adapter.unarmed({token=token,mode=mode,weapon_class="HTH",palette="data/global/Palette/units/pal.dat",direction=3})
	  assert(composite.direction==14 and composite.dcc_direction==3)
	  if token=="NE" then
	    assert(composite.components.S1 and composite.components.S2)
	  end
      local node=render.create("world")
      local frames=node:set_cof_animation(composite.cof,composite.palette,composite.direction,composite.components,"loop",composite.rate)
      assert(frames > 0)
      node:destroy()
    end)
    assert(ok,token..mode..": "..tostring(err))
  end
end
for _,mode in ipairs({"NU","WL","RN"}) do
  local composite=adapter.resolve({token="AM",mode=mode,weapon_class="HTH",palette="data/global/Palette/units/pal.dat",direction=3},{
    active_weapon_set=0,
    items={{container="equipment",slot="rarm",weapon_set=0,weapon_class="1HS",composite={RH="SSD"}}},
  })
  local node=render.create("world")
  assert(node:set_cof_animation(composite.cof,composite.palette,composite.direction,composite.components,"loop",composite.rate) > 0)
  node:destroy()
end
`
	scope := &Scope{}
	if err := runtime.RunScoped(context.Background(), scope, func(state *lua.LState) error { return state.DoString(script) }); err != nil {
		t.Fatal(err)
	}
	if err := scope.Close(); err != nil {
		t.Fatal(err)
	}
	before := capability.Diagnostics().DecodeCalls
	secondScope := &Scope{}
	if err := runtime.RunScoped(context.Background(), secondScope, func(state *lua.LState) error { return state.DoString(script) }); err != nil {
		t.Fatal(err)
	}
	if err := secondScope.Close(); err != nil {
		t.Fatal(err)
	}
	if after := capability.Diagnostics().DecodeCalls; after != before {
		t.Fatalf("warm composite recipes decoded again: before=%d after=%d", before, after)
	}
}

File: internal/content/d2legacy/lua/d2legacy/gameplay/player_composite_test.lua
local test = require("d2legacy.tests/v1")
local fixtures = require("d2legacy.tests.support.fixtures")

local palette = "data/global/Palette/units/pal.dat"

local function install_render_fixture()
    test.mock_module("engine.render/v1", {
        cof_info = function(path)
            test.assert(
                path == "data/global/chars/AM/COF/AMWLHTH.cof"
                    or path == "data/global/chars/AM/COF/AMWL1HS.cof"
                    or path == "data/global/chars/AM/COF/AMRNHTH.cof"
            )
            return {
                directions = 16,
                layers = {
                    { type = "HD", weapon_class = "1ht" },
                    { type = "RA", weapon_class = "hth" },
                    { type = "RH", weapon_class = "hth" },
                },
            }
        end,
        asset_exists = function(path)
            return path == "data/global/chars/AM/HD/AMHDLITWL1HT.dcc"
                or path == "data/global/chars/AM/RA/AMRALITWLHTH.dcc"
        end,
        animdata_info = function(key)
            test.assert(
                key == "AMWLHTH" or key == "AMWL1HS" or key == "AMRNHTH",
                [=[key == "AMWLHTH" or key == "AMWL1HS" or key == "AMRNHTH"]=]
            )
            return { speed = 333, frames = 8, events = {} }
        end,
    }, { "cof_info", "asset_exists", "animdata_info" })
end

local function player_recipe(direction)
    return {
        token = "AM",
        mode = "WL",
        weapon_class = "HTH",
        palette = palette,
        direction = direction,
        class = "Amazon",
        velocitypercent = 0,
        item_fastermovevelocity = 0,
    }
end

local function assert_unarmed_recipe(adapter)
    local composite = adapter.unarmed(player_recipe(3))
    test.assert(
        string.sub(composite.key, 1, 12) == "AM:WL:HTH:14",
        [=[string.sub(composite.key, 1, 12) == "AM:WL:HTH:14"]=]
    )
    test.assert(
        composite.direction == 14 and composite.dcc_direction == 3,
        [=[composite.direction == 14 and composite.dcc_direction == 3]=]
    )
    test.assert(
        composite.components.HD == "data/global/chars/AM/HD/AMHDLITWL1HT.dcc",
        [=[composite.components.HD == "data/global/chars/AM/HD/AMHDLITWL1HT.dcc"]=]
    )
    test.assert(
        composite.components.RA == "data/global/chars/AM/RA/AMRALITWLHTH.dcc",
        [=[composite.components.RA == "data/global/chars/AM/RA/AMRALITWLHTH.dcc"]=]
    )
    test.assert(composite.components.RH == nil, [=[composite.components.RH == nil]=])
    test.assert(composite.rate == 213 and composite.frames == 8, [=[composite.rate == 213 and composite.frames == 8]=])
    test.assert(
        adapter.unarmed(player_recipe(15)).direction == 15,
        [=[adapter.unarmed(player_recipe(15)).direction == 15]=]
    )
end

local function assert_equipped_recipe(adapter)
    local weapon = fixtures.item({
        container = "equipment",
        slot = "rarm",
        weapon_set = 0,
        weapon_class = "1hs",
        composite = { RH = "ssd" },
    })
    local equipment = fixtures.equipment({ items = { weapon } })
    local equipped = adapter.resolve(player_recipe(3), equipment)
    test.assert(
        equipped.cof == "data/global/chars/AM/COF/AMWL1HS.cof",
        [=[equipped.cof == "data/global/chars/AM/COF/AMWL1HS.cof"]=]
    )
    test.assert(
        equipped.components.RH == "data/global/chars/AM/RH/AMRHSSDWLHTH.dcc",
        [=[equipped.components.RH == "data/global/chars/AM/RH/AMRHSSDWLHTH.dcc"]=]
    )
end

local function assert_playback_events(adapter)
    local playback = adapter.new_playback({ mode = "A" })
    local animation = {
        rate = 256,
        frames = 4,
        events = { [2] = 1, [3] = 3 },
        mode = "A",
    }
    local crossed = adapter.advance(playback, animation, 0.09)
    test.assert(playback.frame == 3, [=[playback.frame == 3]=])
    test.assert(#crossed == 2, [=[#crossed == 2]=])
    test.assert(crossed[1].event == 1 and crossed[2].event == 3, [=[crossed[1].event == 1 and crossed[2].event == 3]=])
end

local function assert_network_synchronization(adapter)
    local playback = adapter.new_playback({ mode = "WL" })
    local animation = {
        rate = 256,
        frames = 4,
        events = { [2] = 1, [3] = 3 },
        mode = "WL",
    }
    local initial = adapter.synchronize(playback, animation, 0.09)
    test.assert(playback.frame == 3 and #initial == 0, [=[playback.frame == 3 and #initial == 0]=])
    local crossed = adapter.synchronize(playback, animation, 0.13)
    test.assert(playback.frame == 4 and #crossed == 0, [=[playback.frame == 4 and #crossed == 0]=])
    adapter.synchronize(playback, animation, 0.17)
    test.assert(playback.frame == 1, [=[playback.frame == 1]=])
    adapter.synchronize(playback, animation, 0.01)
    test.assert(playback.frame == 1, [=[playback.frame == 1]=])
end

local function resolves_cof_layer_weapon_classes()
    install_render_fixture()
    local adapter = require("d2legacy.gameplay.player_composite")
    assert_unarmed_recipe(adapter)
    assert_equipped_recipe(adapter)
    assert_playback_events(adapter)
    assert_network_synchronization(adapter)
end

local function movement_stats_not_integrated_distance_drive_animation_playback()
    install_render_fixture()
    local adapter = require("d2legacy.gameplay.player_composite")
    local ordinary = player_recipe(3)
    ordinary.velocity_x = 6
    ordinary.item_fastermovevelocity = 0
    local faster = player_recipe(3)
    faster.velocity_x = 12
    faster.item_fastermovevelocity = 100

    local ordinary_composite = adapter.unarmed(ordinary)
    local faster_composite = adapter.unarmed(faster)
    test.expect(ordinary_composite.rate):equals(213)
    test.expect(faster_composite.rate):equals(340)

    local ordinary_playback = adapter.new_playback(ordinary_composite)
    local faster_playback = adapter.new_playback(faster_composite)
    adapter.advance(ordinary_playback, ordinary_composite, 0.1)
    adapter.advance(faster_playback, faster_composite, 0.1)
    test.assert(
        ordinary_playback.frame ~= faster_playback.frame,
        [=[ordinary_playback.frame ~= faster_playback.frame]=]
    )
    test.expect(ordinary_playback.seconds):equals(faster_playback.seconds)

    local displaced = player_recipe(3)
    displaced.velocity_x = 999
    test.expect(adapter.unarmed(displaced).rate):equals(ordinary_composite.rate)

    local running = player_recipe(3)
    running.mode = "RN"
    test.expect(adapter.unarmed(running).rate):equals(151)

    local retained = adapter.new_playback(ordinary_composite)
    adapter.synchronize(retained, ordinary_composite, 0.02)
    local previous_frame_seconds = 256 / (ordinary_composite.rate * 25)
    local faster_frame_seconds = 256 / (faster_composite.rate * 25)
    local expected_phase = 0.02 / previous_frame_seconds + 0.001 / faster_frame_seconds
    adapter.synchronize(retained, faster_composite, 0.021)
    test.assert(
        math.abs(retained.remainder / faster_frame_seconds - expected_phase) < 0.0000001,
        [=[math.abs(retained.remainder / faster_frame_seconds - expected_phase) < 0.0000001]=]
    )
    test.assert(math.abs(retained.seconds - expected_phase * faster_frame_seconds) < 0.0000001)
    test.expect(retained.authority_seconds):equals(0.021)
end

return test.suite({
    profile = "module",
    tier = "fast",
    covers = { "internal/game/player" },
    cases = {
        test.case("resolves_cof_layer_weapon_classes", {
            { run = resolves_cof_layer_weapon_classes },
        }),
        test.case("movement_stats_not_integrated_distance_drive_animation_playback", {
            { run = movement_stats_not_integrated_distance_drive_animation_playback },
        }),
    },
})
